stratagem-runner/src/stratagem_warning.rs

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) 2019 DDN. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+// Usage: stratagem_warning [FS-NAME|FS-MOUNT-POINT] [NID1] [NID2] ...
+
+extern crate csv;
+#[macro_use]
+extern crate serde_derive;
+
+use libc;
+use liblustreapi;
+use std::env;
+use std::ffi::CStr;
+use std::io;
+use std::process;
+
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "SCREAMING-KEBAB-CASE")]
+struct Record<'a> {
+    path: &'a str,
+    user: &'a str,
+    uid: u32,
+    gid: u32,
+    atime: i64,
+    mtime: i64,
+    ctime: i64,
+    fid: &'a str,
+}
+
+fn write_records(device: &String, args: impl IntoIterator<Item = String>) -> Result<(), io::Error> {
+    let out = io::stdout();
+    let mntpt = match liblustreapi::search_rootpath(&device) {
+        Ok(p) => p,
+        Err(e) => {
+            eprintln!("Failed to find rootpath({}) -> {:?}", device, e);
+            return Err(e);
+        }
+    };
+    let mut wtr = csv::Writer::from_writer(out);
+
+    for fid in args {
+        if let Ok(path) = liblustreapi::fid2path(&device, &fid) {
+            let pb: std::path::PathBuf = [&mntpt, &path].iter().collect();
+            let fullpath = pb.to_str().unwrap();
+            let stat = match liblustreapi::mdc_stat(&fullpath) {
+                Ok(stat) => stat,
+                Err(e) => {
+                    eprintln!("Failed to mdc_stat({}) => {:?}", fullpath, e);
+                    return Err(e);
+                }
+            };
+            let user = unsafe {
```

Comment:
unsafe code should not propagate to this level, it should be encapsulated inside the liblustreapi / liblustreapi-sys crates.

Consumers of those crates should not be required to use any unsafe blocks to interact with them.

Reply:
This is interacting with libc. uid -> username translation.  It doesn't seem to fit in liblustreapi.

Reply:
Ok, How about hoisting it out to a separate fn.

Reply:
ie:

```rust
            fn uid2user(uid: u32) -> Result<String, StratagemError> {
                let pwent = unsafe { libc::getpwuid(uid) };

                if pwent.is_null() {
                    eprintln!("Failed to getpwuid({})", uid);

                    return Err(std::io::Error::from(std::io::ErrorKind::NotFound).into());
                }

                unsafe { Ok(CStr::from_ptr((*pwent).pw_name).to_str()?.to_string()) }
            }
```